src/Functions/array/arrayPrAUC.cpp

```diff
@@ -0,0 +1,241 @@
+  * It represents the probability of the classifier to give positive result if the object has positive label.
+  * Recall = P(score > threshold | label = positive)
+  *
+  * We can draw a curve of values of Precision and Recall with different threshold on [0..1] x [0..1] unit square.
+  * This curve is named "Precision Recall curve" (PR).
+  *
+  * For the curve we can calculate, literally, Area Under the Curve, that will be in the range of [0..1].
+  *
+  * Let's look at the example:
+  * arrayPrAUC([0.1, 0.4, 0.35, 0.8], [0, 0, 1, 1]);
+  *
+  * 1. We have pairs: (-, 0.1), (-, 0.4), (+, 0.35), (+, 0.8)
+  *
+  * 2. Let's sort by score descending: (+, 0.8), (-, 0.4), (+, 0.35), (-, 0.1)
+  *
+  * 3. Let's draw the points:
+  *
+  * threshold = 0.8,  TP = 0, FP = 0, FN = 2, Recall = 0.0, Precision = 1
+  * threshold = 0.4,  TP = 1, FP = 0, FN = 1, Recall = 0.5, Precision = 1
+  * threshold = 0.35, TP = 1, FP = 1, FN = 1, Recall = 0.5, Precision = 0.5
+  * threshold = 0.1,  TP = 2, FP = 1, FN = 0, Recall = 1.0, Precision = 0.666
+  * threshold = 0,    TP = 2, FP = 2, FN = 0, Recall = 1.0, Precision = 0.5
+  *
+  * This implementation uses the right Riemann sum (see https://en.wikipedia.org/wiki/Riemann_sum) to calculate the AUC.
+  * That is, each increment in area is calculated using `(R_n - R_{n-1}) * P_n`,
+  *   where `R_n` is the Recall at the `n`-th point and `P_n` is the Precision at the `n`-th point.
+  *
+  * This implementation is not interpolated and is different from computing the AUC with the trapezoidal rule,
+  *   which uses linear interpolation and can be too optimistic for the Precision Recall AUC metric.
+  */
+
+class FunctionArrayPrAUC : public IFunction
+{
+public:
+    static constexpr auto name = "arrayPrAUC";
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionArrayPrAUC>(); }
+
+private:
+    static Float64
+    apply(const IColumn & scores, const IColumn & labels, ColumnArray::Offset current_offset, ColumnArray::Offset next_offset)
+    {
+        struct ScoreLabel
+        {
+            Float64 score;
+            bool label;
+        };
+
+        size_t size = next_offset - current_offset;
+        PODArrayWithStackMemory<ScoreLabel, 1024> sorted_labels(size);
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            bool label = labels.getFloat64(current_offset + i) > 0;
+            sorted_labels[i].score = scores.getFloat64(current_offset + i);
+            sorted_labels[i].label = label;
+        }
+
+        /// Sorting scores in descending order to traverse the Precision Recall curve from left to right
+        std::sort(sorted_labels.begin(), sorted_labels.end(), [](const auto & lhs, const auto & rhs) { return lhs.score > rhs.score; });
+
+        size_t prev_tp = 0, curr_tp = 0;
+        size_t curr_p = 0;
+
+        Float64 prev_score = sorted_labels[0].score;
+        Float64 curr_precision = 1.0;
+
```

Comment:
I believe that curr_p is current predictions, so I will change the name to curr_pred. If it's not true, please feel free to change it back.